components/remote_settings/src/storage.rs

```diff
@@ -270,6 +295,13 @@ impl Storage {
     }
 }
 
+/// Stores the SQLite connection, which is lazily constructed and can be closed/shutdown.
+enum ConnectionCell {
+    Uninitialized,
+    Initialized(Connection),
+    Closed,
+}
```

Comment:
Lately I've been feeling like this should be the standard way we handle SQLite connections: make them lazily created and give the consumers a way to close them during shutdown.  This simplifies things on the consumer side:
* No need to wrap our components in a lazy wrapper.  You can just construct them with confidence that there's not going to be any IO happening.
* JS code can shutdown components with a `close` method and be sure that all resources are released.  Hopefully this will avoid the late-write errors.

Reply:
that works ok in simple cases, but I'm not sure how that translates to connections which want custom flags? Also slightly concerned this will cause people to create brand new connections for every requests etc rather than managing them well. But the fact this is private for now makes me not worry until it spreads :)

Reply:
Good points.  I won't try to spread this more than it needs to be and we can revisit in the future when we have a concrete use-case.